Mcu update

alatarum opened the pull request:
Bring backplane MCU update implementation

Commit:
storage: update protocol definition for MCU

Some updates in proto

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: report MCU in not functional if can't determine drive type

DriveInterface::Unknown should be threated as internal error since it
only possibly in case of incorrect answer from MCU.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
common: i2c: add methods for raw i2c communication

Since smbus API can't provide functions to send/receive large data blocks
in flexible way, some raw function would be useful.

This commit also adds built-in retry for the case device is not
accessible at the time.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

nest1ing commented on src/storage/backplane_control.cpp:
```diff
@@ -465,3 +496,135 @@ void BackplaneController::hostPowerChanged(bool powered)
+            return;
+        }
+        updater->activation(softwareServer::Activation::Activations::Active);
+    }).detach();
```
Look like memory leak.
Thread once started will never release its own resources.

alatarum replied:
There is no memory leak, detached thread would be cleaned out after completion.

nest1ing commented on src/storage/backplane_control.cpp:
```diff
@@ -465,3 +496,135 @@ void BackplaneController::hostPowerChanged(bool powered)
+                softwareServer::Activation::Activations::Failed);
+            return;
+        }
+        updater->activation(softwareServer::Activation::Activations::Active);
```
This call leads to D-Bus signal sending and should be called from the main thread.

nest1ing reviewed (changes requested):
As I wrote in the internal chat likely the bets solution would move this code to standalone process that will start as a separate `systemd` unit.
This would prevent from a possible race condition between threads and non-thread-safe design of `sdbusplus` library.

Commit:
storage: increase reliability in V0 driver

V0 protocol implementation is very unstable and frequently return
corrupted answer. Because of this a workaround added for some commands
(which are most vulnerable to the problem): application will read same
register several times until we get same answer twice.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: implement backplane MCU update tool

Implment command line tool to update firmware in backplane MCUs. The
tool uses same MCU driver as the storage manager.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: integrate MCU update to storage manager

This change integrate MCU firmware updater to storage manager according
to phosphor software manager architecture.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: fix MCU fw version string len V1

According to protocol description, version string length is 22 bytes,
but actually it could be more then that. Increase string lenght to 32
bytes.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
Cleanup build scripts

Some small fixes

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: add associations to backplane inventory items

Add associations between MCU firmware versions and corresponding
inventory items.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: fix several crashes caused by broken data

Application crashes if it tries to expose to dbus string value which
contains non-printable characters. Couple examples of such crashes:
* Read broken inventory file and try to expose such invalid drive
  information
* Got broken Version string from MCU

To avoid this issues, we should strip any stirng values that we are
going to expose to DBus.

This commit filters out non-printable characters from drive inventory
file and Version string received from MCU.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

nest1ing commented on meson.build:
```diff
@@ -9,7 +9,7 @@ project('obmc-yadro-hardware', 'cpp',
 
 boost = dependency('boost', required: true)
-pdi_dep = dependency('phosphor-dbus-interfaces')
+pdi_dep = dependency('phosphor-dbus-interfaces', required: true)
```
Looks like a double. See line 15.

nest1ing commented on src/storage/backplane_control.cpp:
```diff
@@ -465,3 +497,41 @@ void BackplaneController::hostPowerChanged(bool powered)
+    ss << " -v \"" << imageVersion << "\"";
+    ss << " -B";
+    std::string cmd = ss.str();
+    auto res = std::system(cmd.c_str());
```
This looks weird for me.
1. `-B` can be replaced with `&`. And I dislike both ways as it starts a zombie child. Isn't?
2. `std::system` is blocking call.

I've wrote a simple example of how to start and watch the children using sd-event.
https://github.com/nest1ing/test-fork

nest1ing commented on src/storage/updater.cpp:
```diff
@@ -0,0 +1,346 @@
+    if (background)
+    {
+        pid_t pid;
+        pid = fork();
```
It seems to me an excess as I explain above.

Commit:
move MCU communication code to separate directory

Apart from backplane there are some other components contains MCUs, e.g.
active bezel, which we are going to support later.
Move MCU-specific code from storage manager to separate common
directory.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

Commit:
storage: add status refresh for backplane MCU

Reread information from MCU every 10 seconds to keep data in actual
state.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

nest1ing commented on meson.build:
```diff
@@ -66,6 +65,7 @@ systemd_system_unit_dir = systemd.get_pkgconfig_variable(
 subdir('service_files')
 
 incdir = include_directories('src')
+incdir_mcu = include_directories('src', 'src/mcu')
```
I would prefer to use something like:
```cpp
#include "mcu/backplain_driver.hpp"
```
instead of adding `mcu` to the include directories list.
But it isn't critical and I don't insist on fixing.

nest1ing commented on src/mcu/updater.cpp:
```diff
@@ -0,0 +1,346 @@
+  -d, --dbus-path <object>  Path to software object in dbus. If specified
+                            will manage Activation Result of the object.
+  -p, --progress            Print firmware update progress.
+  -B, --background          Run update in background.
```
This argument initially seemed to me an excess as `command&` is shorter and more familiar than `command -B`.
Now it's unused and can be safely removed.

nest1ing reviewed (dismissed):
Looks good to me except some non-critical cosmetic.

Commit:
swupdate: add tool to work with software packages

Introduce new tool to work with software inages uploaded to the system.
Only 'list' and 'activate' commands implemented at the moment.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

nest1ing commented on src/swupd_activator.cpp:
```diff
@@ -0,0 +1,262 @@
+        auto getProperties = bus.new_method_call(owner.c_str(), path.c_str(),
+                                                 dbus::properties::interface,
+                                                 dbus::properties::getAll);
+        getProperties.append(dbus::software::versionIface);
```
I suggest to not restrict the properties by the version interface.
It allows to receive all of them in the one call.
In this way we lose info about the property interface, but they have unique names and it should be enough.

Commit:
swupdate: more instruments to show version info

Add 'status' command to show currently installed versions.
Add verbose mode to show information about all installation targets.
Cleanup output.

Signed-off-by: Andrei Kartashev <a.kartashev@yadro.com>

nest1ing commented on src/swupd_activator.cpp:
```diff
@@ -0,0 +1,486 @@
+        }
+        if (ver.version.length() <= 35)
+        {
+            fprintf(stdout, "%-24s\t%-35s\t%-10s\t%s\n", verId.c_str(),
```
I still dislike that this line doesn't fit into 80 characters.
`24 + 8(tab) + 35 + 5(tab) + 10 = 82` plus `Target inventory items`, that's  `6(tab) + unlimited string`.
It will look weird via serial console.

nest1ing commented on src/swupd_activator.cpp:
```diff
@@ -0,0 +1,486 @@
+                                               dbus::properties::interface,
+                                               dbus::properties::set);
+        setProperty.append(dbus::software::activationIface,
+                           dbus::software::properties::reqActivation, data);
```
Shouldn't we check this instance has been already activated?
